hcl2template/types.build.go

```diff
@@ -125,3 +132,356 @@ func (p *Parser) decodeBuildConfig(block *hcl.Block) (*BuildBlock, hcl.Diagnosti
+// as VirtualBox, EC2, etc.).
+type CoreHCL2Build struct {
+	BuildName          string
+	Type               string
+	Builder            packer.Builder
+	BuilderConfig      interface{}
+	BuilderType        string
+	hooks              map[string][]packer.Hook
+	Provisioners       []CoreHCL2BuildProvisioner
+	PostProcessors     [][]CoreHCL2BuildPostProcessor
+	CleanupProvisioner CoreHCL2BuildProvisioner
+	TemplatePath       string
+	Variables          map[string]string
+
+	debug         bool
+	force         bool
+	onError       string
+	l             sync.Mutex
+	prepareCalled bool
+
+	cfg *PackerConfig
+}
+
+// CoreHCL2BuildPostProcessor Keeps track of the post-processor and the
+// configuration of the post-processor used within a build.
+type CoreHCL2BuildPostProcessor struct {
+	PostProcessor     packer.PostProcessor
+	PType             string
+	PName             string
+	config            map[string]interface{}
+	keepInputArtifact *bool
+}
+
+// CoreHCL2BuildProvisioner keeps track of the provisioner and the configuration of
+// the provisioner within the build.
+type CoreHCL2BuildProvisioner struct {
+	PType       string
+	PName       string
+	Provisioner packer.Provisioner
+	config      []interface{}
+}
+
+// Returns the name of the build.
+func (b *CoreHCL2Build) Name() string {
+	if b.BuildName != "" {
+		return b.BuildName + "." + b.Type
+	}
+	return b.Type
+}
+
+// The build is prepared previously in PackerConfig.GetBuilds
+// so don't do anything here.
+func (b *CoreHCL2Build) Prepare() (warn []string, err error) {
+	return nil, nil
+}
+
+// Runs the actual build. Prepare must be called prior to running this.
+func (b *CoreHCL2Build) Run(ctx context.Context, originalUi packer.Ui) ([]packer.Artifact, error) {
+	if !b.prepareCalled {
+		panic("Prepare must be called first")
+	}
+
+	// Copy the hooks
+	hooks := make(map[string][]packer.Hook)
+	for hookName, hookList := range b.hooks {
+		hooks[hookName] = make([]packer.Hook, len(hookList))
+		copy(hooks[hookName], hookList)
+	}
+
+	// Add a hook for the provisioners if we have provisioners
+	if len(b.Provisioners) > 0 {
+		hookedProvisioners := make([]*packer.HookedProvisioner, len(b.Provisioners))
+		for i, p := range b.Provisioners {
+			var pConfig interface{}
+			if len(p.config) > 0 {
+				pConfig = p.config[0]
+			}
+			if b.debug {
+				hookedProvisioners[i] = &packer.HookedProvisioner{
+					&packer.DebuggedProvisioner{Provisioner: p.Provisioner},
+					pConfig,
+					p.PType,
+				}
+			} else {
+				hookedProvisioners[i] = &packer.HookedProvisioner{
+					p.Provisioner,
+					pConfig,
+					p.PType,
+				}
+			}
+		}
+
+		if _, ok := hooks[packer.HookProvision]; !ok {
+			hooks[packer.HookProvision] = make([]packer.Hook, 0, 1)
+		}
+
+		hooks[packer.HookProvision] = append(hooks[packer.HookProvision], &packer.ProvisionHook{
+			Provisioners: hookedProvisioners,
+			HCL2Prepare: func(typeName string, data map[string]interface{}) (packer.Provisioner, hcl.Diagnostics) {
```

Comment:
Here's the function that will be called by the ProvisionHook before running each provisioner. 
The function decodes and prepares the provisioner again after setting the build variables with real values.

Reply:
I'm passing a function here, otherwise, there's going to be a cyclic dependency.

Reply:
A cyclic dependency between which packages ?

Reply:
Also, good idea. Could the `CoreBuild` get such a function that does nothing on JSON but re-prepares on HCL2 ?

Reply:
🤔  I think this is possible and would be super nice, we won't need to pass the `PackerConfig`.

Reply:
Nice !